File: python/src/fountain_tools/callback_parser.py
from .fountain import Element
from .parser import FountainParser

class TitleEntry:
    def __init__(self, key, value):
        self.key = key
        self.value = value

class FountainCallbackParser(FountainParser):
    def __init__(self):
        super().__init__()
        # Array of TitleEntry
        self.onTitlePage = None
        # character:string, extension:string, parenthetical:string, line:string, is_dual_dialogue:bool
        self.onDialogue = None
        # text:string
        self.onAction = None
        # text:string, scene_number:string
        self.onSceneHeading = None
        # text:string
        self.onLyrics = None
        # text:string
        self.onTransition = None
        # text:string, level:int
        self.onSection = None
        # text:string
        self.onSynopsis = None
        # no params
        self.onPageBreak = None

        self.ignoreBlanks = True
        self._lastChar = None
        self._lastParen = None

    def add_line(self, line):
        self.mergeActions = False
        self.mergeDialogue = False

        element_count = len(self.script.elements)
        in_title_page = self._inTitlePage

        super().add_line(line)

        if in_title_page and not self._inTitlePage:
            # Finished reading title page
            if self.onTitlePage:
                entries = [TitleEntry(header.key, header.text) for header in self.script.headers]
                self.onTitlePage(entries)

        while element_count < len(self.script.elements):
            self._handle_new_element(self.script.elements[element_count])
            element_count += 1

    def _handle_new_element(self, elem):
        if elem.type == Element.CHARACTER:
            self._lastChar = elem
            return

        if elem.type == Element.PARENTHESIS:
            self._lastParen = elem
            return

        if elem.type == Element.DIALOGUE:
            character =  self._lastChar.name
            extension = self._lastChar.extension
            parenthetical = self._lastParen.text if self._lastParen else None
            line = elem.text
            is_dual_dialogue = self._lastChar.is_dual_dialogue

            self._lastParen = None

            if self.ignoreBlanks and not line.strip():
                return

            if self.onDialogue:
                self.onDialogue(character, extension, parenthetical, line, is_dual_dialogue)
            return

        self._lastChar = None
        self._lastParen = None

        if elem.type == Element.ACTION:
            if self.ignoreBlanks and not elem.text.strip():
                return

            if self.onAction:
                self.onAction(elem.text)
            return

        if elem.type == Element.HEADING:
            if self.ignoreBlanks and not elem.text.strip():
                return

            if self.onSceneHeading:
                self.onSceneHeading(elem.text, elem.scene_number)
            return

        if elem.type == Element.LYRIC:
            if self.ignoreBlanks and not elem.text.strip():
                return

            if self.onLyrics:
                self.onLyrics(elem.text)
            return

        if elem.type == Element.TRANSITION:
            if self.ignoreBlanks and not elem.text.strip():
                return

            if self.onTransition:
                self.onTransition(elem.text)
            return

        if elem.type == Element.SECTION:
            if self.onSection:
                self.onSection(elem.text, elem.level)
            return

        if elem.type == Element.SYNOPSIS:
            if self.onSynopsis:
                self.onSynopsis(elem.text)
            return

        if elem.type == Element.PAGEBREAK:
            if self.onPageBreak:
                self.onPageBreak()
            return

File: csharp/FountainTools/CallbackParser.cs
namespace Fountain;

public class FountainCallbackParser : FountainParser
{
    public class TitleEntry {
        public required string Key;
        public required string Value;
    };

    // List of TitleEntry
    public Action<List<TitleEntry>>? OnTitlePage { get; set; }

     // character:string, extension:string, parenthetical:string, line:string, isDualDialogue:bool
    public Action<string, string?, string?, string, bool>? OnDialogue { get; set; }
    
    // text:string
    public Action<string>? OnAction { get; set; }

    // text:string, sceneNumber:string
    public Action<string, string?>? OnSceneHeading { get; set; }

    // text:string
    public Action<string>? OnLyrics { get; set; }
    
    // text:string
    public Action<string>? OnTransition { get; set; }
    
    // text:string, level:int
    public Action<string, int>? OnSection { get; set; }
    
    // text:string
    public Action<string>? OnSynopsis { get; set; }
    
     // No params 
    public Action? OnPageBreak { get; set; }


    public bool IgnoreBlanks { get; set; } = true;

    private FountainCharacter? _lastChar;
    private FountainParenthesis? _lastParen;

    public FountainCallbackParser() : base()
    {
        _lastChar = null;
        _lastParen = null;
    }

    public override void AddLine(string line)
    {
        MergeActions = false; // Don't merge actions, callbacks need them separated.
        MergeDialogue = false; // Don't merge dialogue, callbacks need them separated.

        int elementCount = Script.Elements.Count;
        bool wasInTitlePage = inTitlePage;

        base.AddLine(line);

        if (wasInTitlePage && !inTitlePage)
        {
            // Finished reading title page
            if (OnTitlePage != null)
            {
                var entries = new List<TitleEntry>();
                foreach (var header in Script.Headers)
                {
                    entries.Add(new TitleEntry{Key = header.Key, Value = header.TextRaw})
;               }
                OnTitlePage(entries);
            }
        }

        while (elementCount < Script.Elements.Count)
        {
            HandleNewElement(Script.Elements[elementCount]);
            elementCount++;
        }
    }

    private void HandleNewElement(FountainElement elem)
    {
        switch (elem.Type)
        {
            case Element.CHARACTER:
                _lastChar = (FountainCharacter)elem;
                break;

            case Element.PARENTHESIS:
                _lastParen = (FountainParenthesis)elem;
                break;

            case Element.DIALOGUE:
                if (_lastChar != null)
                {
                    var character = _lastChar.Name;
                    var extension = _lastChar.Extension;
                    var parenthetical = _lastParen?.TextRaw;
                    var line = elem.TextRaw;
                    var isDualDialogue = _lastChar.IsDualDialogue;
                    
                    _lastParen = null;

                    if (IgnoreBlanks && string.IsNullOrWhiteSpace(line))
                        return;

                    OnDialogue?.Invoke(character, extension, parenthetical, line, isDualDialogue);
                }
                break;

            case Element.ACTION:
                if (IgnoreBlanks && string.IsNullOrWhiteSpace(elem.TextRaw))
                    return;

                OnAction?.Invoke(elem.TextRaw);
                break;

            case Element.HEADING:
                if (IgnoreBlanks && string.IsNullOrWhiteSpace(elem.TextRaw))
                    return;

                var heading = (FountainHeading)elem;
                OnSceneHeading?.Invoke(heading.TextRaw, heading.SceneNumber);
                break;

            case Element.LYRIC:
                if (IgnoreBlanks && string.IsNullOrWhiteSpace(elem.TextRaw))
                    return;

                OnLyrics?.Invoke(elem.TextRaw);
                break;

            case Element.TRANSITION:
                if (IgnoreBlanks && string.IsNullOrWhiteSpace(elem.TextRaw))
                    return;

                OnTransition?.Invoke(elem.TextRaw);
                break;

            case Element.SECTION:
                var section = (FountainSection)elem;
                OnSection?.Invoke(section.TextRaw, section.Level);
                break;

            case Element.SYNOPSIS:
                OnSynopsis?.Invoke(elem.TextRaw);
                break;

            case Element.PAGEBREAK:
                OnPageBreak?.Invoke();
                break;

            default:
                _lastChar = null;
                _lastParen = null;
                break;
        }
    }
}